lib/src/merge.rs

```diff
@@ -664,6 +664,20 @@ where
     /// `FileId` values may differ.
     pub fn with_new_file_ids(&self, file_ids: &Merge<Option<FileId>>) -> Merge<Option<TreeValue>> {
         assert_eq!(self.values.len(), file_ids.values.len());
+        let mut metadata = self
+            .iter()
+            .filter_map(|tree_value| {
+                let Some(TreeValue::File {
+                    executable,
+                    copy_id,
+                    ..
+                }) = borrow_tree_value(tree_value.as_ref())
+                else {
+                    return None;
+                };
+                Some((*executable, copy_id.clone()))
+            })
+            .peekable();
```

Comment:
The defaults can be provided by caller. `write_path_to_store()` knows `copy_id` and `executable`.

```rust
                    // On Windows, we preserve the executable bit from the merged trees.
                    let executable = executable.unwrap_or_else(|| {
                        if let Some(merge) = current_tree_values.to_executable_merge() {
                            conflicts::resolve_file_executable(&merge).unwrap_or(false)
                        } else {
                            false
                        }
                    });
```
(Here, the on-disk `executable` bit should be the last resolved state.)

Other callers can provide dummy value or `None` to make `with_new_file_ids()` panic. I have no idea which is better.

Reply:
Uploaded https://github.com/jj-vcs/jj/pull/8397 with tests for the mentioned 3 cases.